### ISSUE TITLE
Different generated images with the same vector noise depending on # of noises.

### ISSUE THREAD

**Issue by Guim3** [soumith/dcgan.torch#19]
Hi,
I was doing some experiments with the DCGAN with MNIST dataset. Once the GAN is trained, I have noticed that the generated images do not only depend on the original vector noise that originated it, but also the other vector noises that were also given as an input to create other images. 

Let's put a simple example. I have the input vector Z of the generator (9x100x1x1), which is made of 9 subvector noises of dimensionality 1x100x1x1. The generator, then, outputs these nine 32x32 generated images:

![3x3mnist](https://cloud.githubusercontent.com/assets/9153650/16341169/1ade89e8-3a2c-11e6-808a-daebf73a617d.png)

Let's say that, for whatever reason, I'm interested in replicating only the middle left image. So, if I input just the 1x100x1x1 vector instead of the 9x100x1x1, what I obtain instead is this generated image:

![1mnist](https://cloud.githubusercontent.com/assets/9153650/16341256/81bca366-3a2c-11e6-9e34-6b2eb1daaa52.png)

Which is far from being identical to the previous middle left image. So, why is this happening? Shouldn't the generated images be the same regardless of how many input vectors are you using?

This is important if you want to replicate results (for encoding purposes, for example), as you need to exactly input **all** the input vector noises, not only the one you are interested in.

Thanks.

**Comment by soumith** [comment]
this is because of the interactions in batchnorm. Put the model in :evaluate() mode to not have this effect.

**Comment by Guim3** [comment]
Thanks for your quick answer. However, that solved the problem but created a new one. Now I'm able to replicate the results. The problem is that the generated images with evaluate mode are very different to the ones created with training mode:

Training mode
![training](https://cloud.githubusercontent.com/assets/9153650/16342497/ce439356-3a32-11e6-9636-380de1494476.png)

Evaluation mode
![evaluate](https://cloud.githubusercontent.com/assets/9153650/16342512/dbaed0aa-3a32-11e6-943a-17b7fc3294c9.png)

Generated images with evaluation mode are not as close to MNIST as training mode, they have grey background and more bold numbers. 
With training mode activated I assume that batchnorm is using the batch mean and std, while with evaluate mode it uses the mean of all the data, right? May be this is the explanation to this result? Is there any way to avoid this behaviour while being able to replicate results on evaluation time?

Thanks again.

**Comment by soumith** [comment]
i think this is a problem with the running_mean and running_var being a bit off from training.
Before saving the model, do 100 mini-batches of generations (no training) so that they are re-estimated correctly, and then put the model into :evaluate mode and save it.

**Comment by Guim3** [comment]
Thanks once more! I did what you said and now evaluation mode and training mode look the same.
